openmedia/gui/frames.py

```diff
@@ -0,0 +1,193 @@
+
+    # XXX
+    def open_file(self):
+        self.filename = tkFileDialog.askopenfilename()
+        mixer.stop()
+        mixer.init([self.filename])
+        self.f_playlist.l_playlist.clear()
+        self.f_playlist.l_playlist.add_track(os.path.basename(self.filename))
+        self.f_playlist.l_playlist.highlight_next(mixer.song_index)
+        mixer.set_volume(float(self.f_control.s_volume.get()) / 100)
+        self._update_play_icon(self.f_control.buttons[_ControlFrame.PLAY])
+        mixer.play()
+
+class _PlaylistFrame(Tk.Frame):
+    _button_data = [('show_hide', u'≡', False, True),
+                   ('add', u'+', None, False)]
+    _button_names = 'SHOW_HIDE ADD'
+
+    def __init__(self, root):
+        Tk.Frame.__init__(self, root)
+        self.yscroll = Tk.Scrollbar(root)
+        self.buttons = [widgets.PlayerButton(self, text=text,\
+                        name=name, checked=checked, visible=visible) \
+                        for name, text, checked, visible in self._button_data]
+        for index, button in enumerate(self.buttons):
+            if button.visible:
+                button.grid(row=1, column=index)
+        self._create_playlist()
+        self.yscroll.config(command=self.l_playlist.yview)
+
+    def _create_playlist(self):
+        self.l_playlist = widgets.PlayList(self, yscrollcommand=self.yscroll.set)
+        for track in mixer.track_list:
+            self.l_playlist.add_track(os.path.basename(track.get_path()))
+        self.l_playlist.highlight_index(0)
+
+for index, field in enumerate(_PlaylistFrame._button_names.split(' ')):
+    setattr(_PlaylistFrame, field, index)
+
+class _ControlFrame(Tk.Frame):
+    _button_data = [('play_pause', u'▌▌', True),
+                   ('stop', u'■', True),
+                   ('play_next', u'↦', True)]
+    _button_names = 'PLAY STOP NEXT'
+
+    def __init__(self, root):
+        Tk.Frame.__init__(self, root)
+        self.buttons = [widgets.PlayerButton(self, text=text,\
+                        name=name, visible=visible) \
+                        for name, text, visible in self._button_data]
+        self.s_volume = widgets.VolumeSlider(self)
+        for index, button in enumerate(self.buttons):
+            if button.visible:
+                button.grid(row=0, column=index, padx=2)
+        self.s_volume = widgets.VolumeSlider(self)
+        self.s_volume.grid(row=0, column=4, ipady=8)
+        self.s_volume.set(50)
+
+for index, field in enumerate(_ControlFrame._button_names.split(' ')):
+    setattr(_ControlFrame, field, index)
+
+class _StatusFrame(Tk.Frame):
+    def __init__(self, root):
+        pass
+
+class _ProgressFrame(Tk.Frame):
+    def __init__(self, root):
+        Tk.Frame.__init__(self, root)
+        self.s_progress = widgets.PlayerSlider(self, to=mixer.get_song_duration())
+        self.s_progress.grid(row=0, column=0, ipady=10)
+
+    def set_alarm(self, callback):
+        self.s_progress.after(1000, callback)
+
+    def get(self):
```

Comment:
This method could be named something other than simply `get`. Maybe `get_progress`? Or we could leave it out altogether, since `s_progress` is not private.